Add the warning flag to the C/C++ compiler invocation

MisterDA opened the pull request:
See [Sets of warnings for foreign stubs #8597](https://github.com/ocaml/dune/issues/8597).

There are some preliminary commits around naming in the C/C++ compiler handling code, to align the terminology with the upstream OCaml compiler, then a few fixes, then the gist of it.

I add `-Wall` to the C/C++ compiler invocation if GCC/clang are detected, and `-W2` if MSVC or clang-cl are detected. This should help C/C++ OCaml FFI writers to set warnings on by default, which they tend to forget, and that's a *huge* mistake. It is hard to add the right set of warnings in Dune files, see below. If they simply add `-Wall`, they forget about MSVC and clang-cl. Users usually don't know or don't care about these toolchains. Bugs, typos, anything can creep up. Some packages use a discover script with dune-configurator. This is another possible way of forgetting warnings. I believe we should help users catching problems early on by turning on the C/C++ warnings for them.

Note that as long as we don't set the warning as error flag (`-Werror` on cc or `-WX` for msvc), compilation can't fail. No workflow should be broken, the user will just see new warnings reported by the compiler. Users should still be able to disable these flags with `:standard \ -Wall`. 

```
(rule
 (with-stdout-to c_flags.cc (echo "(-Wall)")))

(rule
 (with-stdout-to c_flags.msvc (echo "(/W2)")))

(library
 (name mylib)
 (foreign_stubs
  (language c)
  (flags :standard (:include c_flags.%{ocaml-config:ccomp_type}))
  (names mystubs)))
```

This PR is best reviewed commit by commit.

cc @voodoos @rgrinberg @nojb

MisterDA committed:
Fix detection of clang-cl in cc_vendor

clang-cl is a drop-in replacement for MSVC, and exposes the same set
of flags as MSVC. It defines both `_MSC_VER` and `__clang__`, but we
want it do have precedence over clang, just as clang (on Unix) defines
both `__GNUC__` and `__clang__`, which is why we want it to have
precedence over GCC.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
Disable MSVC copyright output when discovering the cc_vendor

MSVC has an annoying ‘feature’: it prints a copyright statement
everytime it's called. Fortunately, it can be silenced by passing the
`/nologo` argument. MSVC always prints the name of the file it's
compiling.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
Improve warning message in C/C++ compiler detection

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
Differentiate the ccomp_type and the cc_vendor

ccomp_type is given by the OCaml compiler. It's either 'cc' or 'msvc',
but could be something else in future versions. The C compiler vendor
is another specialization. This aligns with the OCaml compiler
terminology.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
Rename Cxx_rules and Cxx_flags to Cc_rules and Cc_flags

They describe how to call the C/C++ compiler and are not directly tied
to C++ anymore.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA committed:
Add a warning flag to the C/C++ compiler invocation

Note that as long as we don't set the warning as error flag (-Werror
on cc or -WX for msvc), compilation can't fail. No workflow will be
broken, the user will just see new warnings reported by the compiler.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

nojb commented:
I am roughly in favour and I agree that more warnings is better than less warnings. However, certain users will probably be annoyed by the change (I am thinking of certain warnings that are a bit complicated to "fix", eg `alloc-size-larger-than`; see eg https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85783#c3).

I will try to take a look at the patch later in the week.

nojb reviewed (commented):
LGTM

I think the change should be versioned though (@rgrinberg: do you confirm?)

MisterDA committed:
test: rename some variables for consistency

echo-ing the variables is somewhat brittle but removes the spaces
inbetween flags.

Signed-off-by: Antonin Décimo <antonin@tarides.com>

MisterDA commented:
> I think the change should be versioned though (@rgrinberg: do you confirm?)

Did we version when we introduced `-fdiagnostics-color`? If we have to version it'll take *ages* before this change reaches the devs. Lots of library authors are reluctant to use the latest Dune lang version, as it forces rebuilds downstream, or an unwanted bound of their users.

MisterDA commented:
Maybe we could go further and enable the `-Werror` _Make all warnings into errors._ flags in debug mode (`-WX` for msvc). Of course, *never* in release mode (the rationale being that new compilers might introduce new warnings, that old code has no way of fixing). It could also follow `-warn-error -A`, but that's above my knowledge of Dune.

nojb commented:
> Maybe we could go further and enable the `-Werror` _Make all warnings into errors_ flag in debug mode (`-WX` for msvc).

I would be very careful about this change. We did that with OCaml warnings, and it is a recurring pain point by users. So I would leave this for a later PR at least.

MisterDA commented:
Should I ask the community about this? Should I break the refactors out of this PR?

nojb commented:
> Should I ask the community about this?

About making the warnings fatal in `dev` mode? Sure, but you should also discuss that more widely with the maintainers of Dune (maybe during the weekly dev meeting); I don't think it is the kind of change that should be done lightly.

> Should I break the refactors out of this PR?

I already reviewed the PR, but if it is not too much work, indeed having the refactorings and the semantic changes in separate PRs would be preferrable. It makes it easier to review and to be confident of the correctness of the change.

Having a second approval would be nice. @Alizter: are you able to take a look?

MisterDA commented:
> > Should I ask the community about this?
> 
> About making the warnings fatal in `dev` mode? Sure, but you should also discuss that more widely with the maintainers of Dune (maybe during the weekly dev meeting); I don't think it is the kind of change that should be done lightly.

Sorry, that was unclear. I agree with you not to make the warnings fatal. I was thinking of asking the community about adding warning flags.

Alizter reviewed (commented):
In `test/blackbox-tests/test-cases/watching/github9225.t` I mentioned `ccomp` rules, we should replace that with `cc_vendor` since that is the new name.

In `test/expect-tests/findlib_tests.ml` we have a dummy `Lib_config.t`. I believe that this should now be `Cc`.

Did the `ccomp_type` value in `ocamlc -config` ever output anything other than `msvc` or `cc`?

Alizter commented on src/dune_rules/cc_rules.ml:
```diff
@@ -35,8 +33,8 @@ let rules ~sctx ~dir =
          [ (match Ocaml_config.ccomp_type ocfg with
-            | Msvc -> A "/EP"
-            | Other _ -> As [ "-E"; "-P" ])
+            | Msvc -> As [ "/nologo"; "/EP" ]
```
I also noticed we are now passing `/nologo` to msvc. Looking at the documentation this appears to be harmless, but it might be worth pointing out to others if they ever come back to this PR.

MisterDA replied:
This is the content of the commit message:

> Disable MSVC copyright output when discovering the cc_vendor
>
> MSVC has an annoying ‘feature’: it prints a copyright statement
everytime it's called. Fortunately, it can be silenced by passing the
`/nologo` argument. MSVC always prints the name of the file it's
compiling.

MisterDA commented on test/blackbox-tests/test-cases/foreign-stubs/c-flags.t/run.t:
```diff
@@ -5,11 +5,11 @@ In the following tests, foo.c is built with the :standard set of flags while
-  $ O_CCPPF=$(echo $O_CCPPF | sed -e 's/^[ \t]*//')
+  $ O_CFLAGS=$(ocamlc -config-var ocamlc_cflags)
+  $ O_CPPFLAGS=$(ocamlc -config-var ocamlc_cppflags)
+  $ O_CC=$(echo $O_CC | sed -e 's/^[[:space:]]*//')
```
Done, thanks!

MisterDA commented on test/blackbox-tests/test-cases/cxx-flags.t/run.t:
```diff
@@ -131,7 +131,7 @@ ccomp is not computed if not required
   OK
 
-  $ [ -f _build/default/.dune/ccomp/ccomp ]
+  $ [ -f _build/default/.dune/cc_vendor/cc_vendor ]
```
Done, thanks!

MisterDA commented:
Thanks for the reviews!

> In `test/blackbox-tests/test-cases/watching/github9225.t` I mentioned `ccomp` rules, we should replace that with `cc_vendor` since that is the new name.

Done, thanks!

> In `test/expect-tests/findlib_tests.ml` we have a dummy `Lib_config.t`. I believe that this should now be `Cc`.

Thanks, done!
 
> Did the `ccomp_type` value in `ocamlc -config` ever output anything other than `msvc` or `cc`?

No.

Alizter reviewed (approved):
LGTM thanks!

nojb commented:
Thanks for the prompt review @Alizter!
